src/frontend/src/lib/components/send/QRCodeModal.svelte

```diff
@@ -0,0 +1,85 @@
+<script lang="ts">
+	import { i18n } from '$lib/stores/i18n.store';
+	import {
+		QRCodeReaderModal,
+		WizardModal,
+		type WizardStep,
+		type WizardSteps
+	} from '@dfinity/gix-components';
+	import { type Token } from '@dfinity/utils';
+	import { toastsError } from '$lib/stores/toasts.store';
+	import type { QrResponse, QrStatus } from '$lib/types/qr-code';
+	import { decodeQrCode } from '$lib/utils/qr-code.utils';
+
+	export let steps: WizardSteps;
+	export let currentStep: WizardStep | undefined = undefined;
+
+	const STEP_QRCODE = 'QRCode';
+
+	let stepsPlusQr: WizardSteps;
+	$: stepsPlusQr = [
+		...steps,
+		{
+			name: STEP_QRCODE,
+			title: $i18n.send.text.scan_qr
+		}
+	];
+
+	let modal: WizardModal;
+
+	const goToStep = (stepName: string) => {
+		const stepNumber = stepsPlusQr.findIndex(({ name }) => name === stepName);
+		modal.set(stepNumber);
+	};
+
+	let resolveQrCodePromise:
+		| (({ status, code }: { status: QrStatus; code?: string }) => void)
+		| undefined = undefined;
+
+	export const scanQrCode = async ({
+		expectedToken
+	}: {
+		expectedToken: Token;
+	}): Promise<QrResponse> => {
+		const prevStep = currentStep;
+		goToStep(STEP_QRCODE);
+
+		return new Promise<{ status: QrStatus; code?: string | undefined }>((resolve) => {
+			resolveQrCodePromise = resolve;
+		})
+			.then(({ status, code }) => {
+				if (status === 'token_incompatible') {
+					toastsError({
+						msg: { text: $i18n.send.error.incompatible_token }
+					});
+					return Promise.reject(new Error('Token incompatible'));
+				}
+				return decodeQrCode({ status, code, expectedToken });
```

Comment:
Just thought at an additional question: shouldn't be this a property as well? i.e. can we decode Ethereum and IC QR code with the same decoder?

Reply:
Yes, agreed, it is better.

So, I double-checked before writing it down and for token/destination/amount it is the same schema for both ETH and IC, but for ETH and ERC20 they have more optional parameter, that we could at some point use too (for example gasLimit).

Furthermore, for ERC20, actually I need to extract the token address from the QR code.

Reply:
One thing I should mention is that the ICRC standard for IC qr codes will most probably slightly differs - i.e. the current implementation is more "a placeholder" kind of than the specification. So ultimately we most probably gonna need different decoder.

But, we can do that later for sure.

Here the discussed spec: https://github.com/dfinity/ICRC/issues/22

Reply:
I am creating a new PR #1346 for that